pkg/decoder/decoder.go

```diff
@@ -7,102 +7,247 @@ package decoder
+func (d *Decoder) Stop() {
+	close(d.InputChan)
+}
+
 // run lets the Decoder handle data coming from the InputChan
 func (d *Decoder) run() {
 	for data := range d.InputChan {
-		d.decodeIncomingData(data.content, data.offset)
+		if d.isMultilineEnabled() {
+			d.ackIncomingData()
+		}
+		endsWithEOL, offset := d.decodeIncomingData(data.content, data.offset)
+		if endsWithEOL && d.isMultilineEnabled() {
+			// when the data ends with `\n`, it may be the end of the multi-line message
+			d.ackEndIncomingData(offset)
+		}
 	}
 	d.OutputChan <- message.NewStopMessage()
 }
 
-// Stop stops the Decoder
-func (d *Decoder) Stop() {
-	close(d.InputChan)
+// ackIncomingData stops the inner timer that flushes the multi-line buffer
+// as we got new data
+func (d *Decoder) ackIncomingData() {
+	if d.timer != nil {
+		d.timer.Stop()
+	}
 }
 
-var truncatedMsg = []byte("...TRUNCATED...")
-var truncatedLen = len(truncatedMsg)
-var maxMessageLen = config.MaxMessageLen - truncatedLen
+// ackEndIncomingData starts the timer that sends the last line when fired
+func (d *Decoder) ackEndIncomingData(offset int64) {
+	d.timer = time.AfterFunc(d.countDownDuration, func() {
+		// FIXME do sometyhinh
+		newLine := make([]byte, d.lineBuffer.Len())
+		copy(newLine, d.lineBuffer.Bytes())
+		defer d.lineBuffer.Reset()
+		d.msgBuffer.Write(newLine)
+		d.sendBufferedMessage(offset)
+	})
+}
 
-// sendBuffuredMessage flushes the buffer and sends the message
-func (d *Decoder) sendBuffuredMessage(offset int64) {
+// sendBufferedMessage flushes the msgBuffer and sends the message
+// FIXME name
+func (d *Decoder) sendBufferedMessage(offset int64) {
 	msg := make([]byte, d.msgBuffer.Len())
 	// d.msgBuffer.Bytes() returns a slice to the []byte, we thus need to copy it
 	copy(msg, d.msgBuffer.Bytes())
+	defer d.msgBuffer.Reset()
 	if len(msg) > 0 {
 		m := message.NewMessage(msg)
 		o := message.NewOrigin()
 		o.Offset = offset
 		m.SetOrigin(o)
 		d.OutputChan <- m
 	}
-	d.msgBuffer.Reset()
 }
 
-// decodeIncomingData splits raw data based on `\n`, creates and sends messages to a channel
-func (d *Decoder) decodeIncomingData(inBuf []byte, offset int64) {
+// processBufferedLine checks the new line, appends the whole line or just a piece to the message
+// and sends the message if needed
+func (d *Decoder) processNewLine(offset int64) {
+	newLine := make([]byte, d.lineBuffer.Len())
+	copy(newLine, d.lineBuffer.Bytes()) // FIXME: do we need to copy
+	defer d.lineBuffer.Reset()
+
+	if !d.isMultilineEnabled() {
+		// FIXME pass directly without copying
+		d.msgBuffer.Write(newLine)
+		d.sendBufferedMessage(offset)
+		return
+	}
+
+	// first line of multi line message: write
+	if d.msgBuffer.Len() == 0 {
+		d.msgBuffer.Write(newLine)
+		return
+	}
+	// FIXME else?
+	// we already have a beg of multi line message
+	if d.newLineRe.Match(newLine) {
+		// we got a new line, send previous line
+		d.sendBufferedMessage(offset - int64(1+len(newLine))) // remove '\n' + new line length
```

Comment:
How would this behave on windows where the default line break is `\r\n`, not `\n` ?